go/vt/topo/shard.go

```diff
@@ -616,6 +614,73 @@ func (ts *Server) FindAllTabletAliasesInShardByCell(ctx context.Context, keyspac
 	return result, err
 }
 
+// GetTabletsByShard returns the tablets in the given shard using all cells.
+// It can return ErrPartialResult if it couldn't read all the cells, or all
+// the individual tablets, in which case the result is valid, but partial.
+func (ts *Server) GetTabletsByShard(ctx context.Context, keyspace, shard string) ([]*TabletInfo, error) {
+	return ts.GetTabletsByShardCell(ctx, keyspace, shard, nil)
+}
+
+// GetTabletsByShardCell returns the tablets in the given shard. It can return
+// ErrPartialResult if it couldn't read all the cells, or all the individual
+// tablets, in which case the result is valid, but partial.
+func (ts *Server) GetTabletsByShardCell(ctx context.Context, keyspace, shard string, cells []string) ([]*TabletInfo, error) {
+	span, ctx := trace.NewSpan(ctx, "topo.GetTabletsByShardCell")
+	span.Annotate("keyspace", keyspace)
+	span.Annotate("shard", shard)
+	span.Annotate("num_cells", len(cells))
+	defer span.Finish()
+	var err error
+
+	if len(cells) == 0 {
+		cells, err = ts.GetCellInfoNames(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if len(cells) == 0 { // Nothing to do
+			return nil, nil
+		}
+	}
+
+	// Divide the concurrency limit by the number of cells. If there are more
+	// cells than the limit, default to concurrency of 1.
+	cellConcurrency := 1
+	if len(cells) < DefaultConcurrency {
+		cellConcurrency = DefaultConcurrency / len(cells)
+	}
+
+	mu := sync.Mutex{}
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(DefaultConcurrency)
+
+	tablets := make([]*TabletInfo, 0, len(cells))
+	options := &GetTabletsByCellOptions{
+		Concurrency: cellConcurrency,
+		KeyspaceShard: &KeyspaceShard{
+			Keyspace: keyspace,
+			Shard:    shard,
+		},
+	}
```

Comment:
I think the common path will have keyspace and shard both as empty strings? If so, we could pass a nil KeyspaceShard in those cases to make the work done in GetTabletsByCell a little easier:
```
	tablets := make([]*TabletInfo, 0, len(cells))
	var kss *KeyspaceShard
	if keyspace != "" || shard != "" {
		kss = &KeyspaceShard{
			Keyspace: keyspace,
			Shard:    shard,
		}
	}
	options := &GetTabletsByCellOptions{
		Concurrency:   cellConcurrency,
		KeyspaceShard: kss,
	}
```

Reply:
Good idea, updated